runtime/lua/vim/_system.lua

```diff
@@ -63,10 +64,29 @@ local function new_systemobj(state)
   }, { __index = SystemObj })
 end
 
---- @param signal integer
+local PROCESS_KILL_WAIT = 5000
+
+--- @param signal integer|string
 function SystemObj:kill(signal)
   local state = self._state
   state.handle:kill(signal)
+
+  -- wait for the process not to exist
```

Comment:
This seems to assume that any signal will result in terminating the process which may not always be true, especially for things like SIGINFO or SIGUSR1. Some processes may even need multiple SIGINTs to shut down in some cases or ignore them entirely.

Reply:
Yeah I realised this too after I posted this.

The only thing I can think of is to create a luv check object to see if the pipes need closing on each event loop iteration in the case the process has terminated.

Reply:
Actually, based on some additional print statements it looks like the existing `close_handles()` call in the `on_exit` callback is already covering this, so maybe it's safe to remove the call in `:kill()`?

These are the changes I was testing with:
```diff
diff --git a/runtime/lua/vim/_system.lua b/runtime/lua/vim/_system.lua
index 6f5e95eb2..05a27a9a2 100644
--- a/runtime/lua/vim/_system.lua
+++ b/runtime/lua/vim/_system.lua
@@ -67,7 +67,6 @@ end
 function SystemObj:kill(signal)
   local state = self._state
   state.handle:kill(signal)
-  close_handles(state)
 end
 
 local MAX_TIMEOUT = 2 ^ 31
@@ -267,7 +266,9 @@ function M.run(cmd, opts, on_exit)
     detached = opts.detach,
     hide = true,
   }, function(code, signal)
+    print('in on_exit')
     close_handles(state)
+    print('handles closed')
     if state.timer then
       state.timer:stop()
       state.timer:close()
```

Reply:
Yep, looks like `on_exit` is called when the process eventually finishes.